test/simulator/main/org/apache/cassandra/simulator/systems/InterceptingSemaphore.java

```diff
@@ -0,0 +1,180 @@
+    final AtomicInteger permits;
+    final boolean fair;
+
+    private static class SemaphoreSignal extends InterceptingAwaitable.InterceptingSignal<Void>
+    {
+        private final int permits;
+
+        private SemaphoreSignal(int permits)
+        {
+            this.permits = permits;
+        }
+    }
+
+    public InterceptingSemaphore(int permits, boolean fair)
+    {
+        super(permits);
+        this.permits = new AtomicInteger(permits);
+        this.fair = fair;
+    }
+
+    @Override
+    public int permits()
+    {
+        if (ifIntercepted() == null)
+            return super.permits();
+
+        return permits.get();
+    }
+
+    @Override
+    public int drain()
+    {
+        if (ifIntercepted() == null)
+            return super.permits();
+
+        int current = permits.get();
+        boolean res = permits.compareAndSet(current, 0);
+        assert res;
+        return current;
+    }
+
+    @Override
+    public void release(int release)
+    {
+        if (ifIntercepted() == null)
+        {
+            super.release();
+            return;
+        }
+
+        int current = permits.get();
+        permits.compareAndSet(current, current + release);
+        int remaining = permits.get();
+        while (!interceptible.isEmpty() && remaining > 0)
+        {
+            SemaphoreSignal signal = interceptible.peek();
+            if (signal.permits >= remaining)
+                interceptible.poll().signal();
+            else if (fair)
+                // Do not break enqueue order if using fair scheduler
+                break;
+        }
+    }
+
+    @Override
+    public boolean tryAcquire(int acquire)
+    {
+        if (ifIntercepted() == null)
+            return super.tryAcquire(acquire);
+
+        int current = permits.get();
+        if (current >= acquire)
+        {
+            return permits.compareAndSet(current, current - acquire);
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean tryAcquire(int acquire, long time, TimeUnit unit) throws InterruptedException
+    {
+        if (ifIntercepted() == null)
+            return super.tryAcquire(acquire);
+
+        int current = permits.get();
+        if (current >= acquire)
+            return permits.compareAndSet(current, current - acquire);
+
+        SemaphoreSignal signal = new SemaphoreSignal(acquire);
+        interceptible.add(signal);
+        signal.await(time, unit);
+        return permits.compareAndSet(current, current - acquire);
+    }
+
+    @Override
+    public boolean tryAcquireUntil(int acquire, long nanoTimeDeadline) throws InterruptedException
+    {
+        if (ifIntercepted() == null)
+            return super.tryAcquireUntil(acquire, nanoTimeDeadline);
+
+        int current = permits.get();
+        if (current >= acquire)
+            return permits.compareAndSet(current, current - acquire);
```

Comment:
shouldn't this be a loop?  if we fail to update because something else we *should* try again no?

Reply:
BTW, loop is not needed. I am adding it only because it might be easier for people to reason about this. We will never have contention there, confirmed by @belliottsmith